ONSdigital/dp-import-cantabular-dataset
# Feature/dont consume on unhealthy

### What

Stop consuming Kafka messages on unhealthy state, and start again on healthy state. Trello card: https://trello.com/c/NxIax7FQ/5349-start-stop-consuming-on-healthy-unhealthy-for-dp-import-cantabular-dataset-8

- dp-healthcheck has been upgraded to version 1.2.1, which implements a subscirbe/notify mechanism for health states (code level subscribes)
- dp-kafka has been upgraded to version 3, which implements the health callback - starting consuming messages on healthy state, and stop consuming messages on unhealthy state.
- kafka consumer is subscribed to the necessary Checks on health check
- Added feature flag to enable/disable this behaviour - if the behavior is disabled, start the kafka consumer straight away instead of subscribing it to health updates.
- removed consumer loop and related code/structs, and use the new handler register functionality implemented in dp-kafka (simplifying the code substantially in the service)
- Upgrade log.go to version 2

Also includes kafka health state fix: https://trello.com/c/NmmQVJGg/5435-kafka-client-fails-health-check-when-msk-security-patches-are-applied-s1

### How to review

- Make sure unit/component tests pass
- Make sure code changes make sense
- (info) tested locally with docker-compose:
  - Kill a dependency container and check that this service stops consuming from kafka as soon as the healthcheck is executed (log message: `kafka consumer group has finished consuming: sarama consumer group session cleanup finished: all go-routines have completed`)
  - Start the container again and check that this service starts consumign from kafka again as soon as the healthcheck is executed (log message: `kafka consumer group is consuming: sarama consumer group session setup ok: a new go-routine will be created for each partition assigned to this consumer`)
- (info) regression test with docker-compose: run a full cantabular import with this branch checked out, and validate that it is still working.
- (info) tested that after killing one Kafka broker, the `/health` endpoint returns WARNING state for the consumer and producer. Killing 2 brokers makes the producer health to be CRITICAL.

### Who can review

Anyone
